### PR TITLE
Feature/notebook

### LINKED ISSUE #1824: Clojure Notebooks
My initial idea is to have a “open this clojure file as a notebook” context menu and have every top-level expression be a code cell and every comment markdown cell. Mostly the same stuff clerk is already doing. The advantage of having this inside Calva would be interaction with all the other Calva stuff and Calva itself would gain a rich inline eval output space.

So my vision is less of a data science notebook that you pass around to other people, although that’s still possible, and more of a deeper integration of eval stuff. Basically mini portal inline windows or similar. If I could just do that in the normal text view, I’d keep it there, but we’re not allowed to have full html in there.

### PR DESCRIPTION
**Pull request by Cyrik**
## What has changed?

- adds a notebook provider
- adds a Notebook Kernel
You can try it out by right clicking on any clojure file -> `Open with...` -> `Calva Notebook`.

Closes #1824 

## My Calva PR Checklist

If this PR involves code changes, I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Directed this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I have changed the PR base branch, so that it is not `published`. (Sorry for the nagging.)
- [x] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.
- [x] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and either tested it there if so, or mentioned it in the PR.
- [x] Added to or updated docs in this branch, if appropriate
- [x] Tests
  - [x] Tested the particular change
  - [x] Figured if the change might have some side effects and tested those as well.
  - [x] Smoke tested the extension as such.
  - [x] Tested the VSIX built from the PR (so, after you've submitted the PR). You'll find the artifacts by clicking _Show all checks_ in the CI section of the PR page, and then _Details_ on the `ci/circleci: build` test.
- [x] Referenced the issue I am fixing/addressing _in a commit message for the pull request_.
  - [x] If I am fixing the issue, I have used [GitHub's fixes/closes syntax](https://help.github.com/en/articles/closing-issues-using-keywords)
  - [x] If I am fixing just part of the issue, I have just referenced it w/o any of the "fixes” keywords.
- [x] Created the issue I am fixing/addressing, if it was not present.
- [x] Formatted all JavaScript and TypeScript code that was changed. (use the [prettier extension](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) or run `npm run prettier-format`)
- [x] Confirmed that there are no linter warnings or errors (use the [eslint extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint), run `npm run eslint` before creating your PR, or run `npm run eslint-watch` to eslint as you go).

Ping @pez, @bpringe, @corasaurus-hex, @Cyrik

### CODE REVIEW

**Comment by Cyrik** [comment]
So this is working "fully".
There's only one thing I'd like to fix before releasing this as MVP: saving currently reformats your file a little since I don't keep track of the whitespace between code blocks. @PEZ I'm using the LispCursor to get the top-level forms, is there something better that would also give me the whitespace ranges? Otherwise, I'll build it myself.
Other than that I'm being very unsafe with error handling and pretty printing, but I think that's ok for now, especially since this is a different view on the file that you can ignore if it's not working for your particular output size.

The next steps would be to use the configured pretty printer and to get all the usual tooling running, like autocomplete and stuff. Only a tiny bit of that is actually something we'll have to do in Calva, most of it is going to happen in clojure-lsp.

**Comment by PEZ** [comment]
Thanks for this work, @Cyrik! ❤️ 

I'm unfamiliar with notebooks. Can you explain a bit about what the provider and kernel responsibilities are?

> I'm using the LispCursor to get the top-level forms, is there something better that would also give me the whitespace ranges? Otherwise, I'll build it myself.

Not really. There are some facilities for collecting whitespace information around the cursor, but I don't know how much use you could have for those. If _whitespace_ here means _everything else, but top level forms_, i.e. including comments, then the whitespace ranges are something like the inverse of the top level ranges, I think. Otherwise... if comments should be editable, maybe we should have a version of the top-level-ranges that includes ranges for the comment blocks?

I'd like to merge this, as is, and then we can build on it. However, would be good with a docs page for Notebooks so that users can orient themselves a bit on what to expect.

**Comment by Cyrik** [comment]
Sure, but I'll have to keep it brief since I have to run for now.
The notebookProvider is a class that's gets called by vscode when you open or save a notebook. The deserialize method gets the content of the opened file and is supposed to return notebook cells that represent the file. The serialize method gets called on save and gets the cells and need to transform them into a string that is going to be written to the file.
You can go more complicated here and load arbitrary stuff from other sources or do transformations, but my basic implementation just wants to make every top level sexp into its own code block. To perserve comments and whitespace I'll have to go a little extra so that I can write them back out on save.

The inverse sounds like an OK way to go for now, I'll try to add that this to the PR. I'll also add a little bit to the docs. As you said, we can make this better in follow-up PRs and this feature is opt-in anyway.

**Comment by Cyrik** [comment]
Fixed saving of whitespace/comments.
Handling of this could be a lot better, but it will save out the same input, so I'm happy for an alpha.

Also added a doc, although I'm not sure what to put into it 😅 

Happy for any suggestions for v0, otherwise I'd add a changelog entry and call this done, and work on improvements in the next version.

File: package-lock.json [hunk 1]
```diff
@@ -1294,6 +1294,7 @@
       "version": "1.1.0",
       "resolved": "https://registry.npmjs.org/@npmcli/fs/-/fs-1.1.0.tgz",
       "integrity": "sha512-VhP1qZLXcrXRIaPoqb4YA55JQxLNF3jNR4T55IdOJa3+IFJKNYHtPvtXx8slmeMavj37vCzCfrqQM1vWLsYKLA==",
+      "deprecated": "this version had an improper engines field added, update to 1.1.1",
```
Comment (bpringe) [inline package-lock.json:1297]: yikes... I'll make a task for this

Comment (bpringe) [reply package-lock.json:1297]: And the other "deprecated" dep issues in this file

Comment (bpringe) [reply package-lock.json:1297]: https://github.com/BetterThanTomorrow/calva/issues/1822

File: test-data/notebook.clj [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+   :deeper {:a 1, "foo" :bar, [1 2 3] (vec (range 2000))}})
+
+(defn foo []
+  (println "bar"))
+
+(foo)
+
+hover-map
+
+"<html> <body>testing<h2>HTML Image</h2><img src='https://raw.githubusercontent.com/BetterThanTomorrow/calva/dev/assets/calva-64h.png'></body></html>"
```
Comment (bpringe) [inline test-data/notebook.clj]: Prime opportunity for a Rickroll.

**Review by bpringe — APPROVED** [review]

**Review by bpringe — CHANGES_REQUESTED** [review]
Just need a changelog entry and I think we're good to merge. 🎉 Thanks again for this!

**Comment by PEZ** [comment]
Been playing a bit with this now and I must say it's lovely. ❤️ This could very well get me started as a notebooks user.

I've updated the documentation page a bit and added it to the **Features** section.

**NB**: There is a quite nasty bug with comments preceding the first form not being retained. 

Some suggestions:

* I think the name **Clojure Notebook** is better than **Calva Notebook**. I used this name in the docs update, but haven't changed it in the config/code.
* I'd like an issue created about adding basic Notebook support. It could include a bit about what notebooks are and why it would be good to have them. This issue is what should be referred in the change log.
* Related to the issue ticket, I think we should start a Discussions thread about Notebooks. It seems to be a very large subject with lots of choices and priorities and trade-offs.

File: test-data/notebook.clj (hunk 2 above)
Comment (Cyrik) [reply test-data/notebook.clj]: done

**Review by PEZ — COMMENTED** [review]
See my comments above. =)

**Comment by Cyrik** [comment]
I've renamed the description to "Clojure Notebook" but set the id to "calva-clojure-notebook" to prevent clashes with a potential extension that also provides Clojure notebooks.
Issue and discussion are also created and the "comment or whitespace before the first exp" is also fixed.
I think this can be merged @PEZ

File: docs/site/notebooks.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+!!! Warning "WIP: Notebook support is very basic and experimental"
+   * You might experience loss of file contents when saving a Clojure Notebook. _Have backups of any files you open as notebooks._
+   * Tooling around autocomplete, go to def, linting and such does not work at all yet.
+   * There is no Markdown support.
+   * Etcetera.
+
+   Please help test the feature. We're looking forward to your feedback!
+
+You can open any clojure file as a notebook by right clicking the file -> `Open with...` -> `Calva Notebook`.
```
Comment (bpringe) [inline docs/site/notebooks.md]:
```suggestion
You can open any Clojure file as a notebook by right clicking the file -> `Open with...` -> `Clojure Notebook`.
```

Comment (bpringe) [reply docs/site/notebooks.md]: I think this should be updated, right?

**Review by bpringe — APPROVED** [review]
Approving, but I made a doc update suggestion.

**Pull request merged**